app/code/Magento/CatalogUrlRewrite/Model/Product/Validator.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\CatalogUrlRewrite\Model\Product;
+
+use Magento\Catalog\Model\Product;
+use Magento\CatalogUrlRewrite\Model\ProductUrlPathGenerator;
+use Magento\UrlRewrite\Model\UrlFinderInterface;
+use Magento\UrlRewrite\Service\V1\Data\UrlRewrite;
+use Magento\CatalogUrlRewrite\Model\ProductUrlRewriteGenerator;
+use Magento\Store\Model\StoreManagerInterface;
+
+/**
+ * Url Rewrites Product validator.
+ */
+class Validator
+{
+    /**
+     * @var ProductUrlPathGenerator
+     */
+    private $productUrlPathGenerator;
+
+    /**
+     * @var UrlFinderInterface
+     */
+    private $urlFinder;
+
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * @param ProductUrlPathGenerator $productUrlPathGenerator
+     * @param UrlFinderInterface $urlFinder
+     * @param StoreManagerInterface $storeManager
+     */
+    public function __construct(
+        ProductUrlPathGenerator $productUrlPathGenerator,
+        UrlFinderInterface $urlFinder,
+        StoreManagerInterface $storeManager
+    ) {
+        $this->productUrlPathGenerator = $productUrlPathGenerator;
+        $this->urlFinder = $urlFinder;
+        $this->storeManager = $storeManager;
+    }
+
+    /**
+     * Find Url Key conflicts of a product.
+     *
+     * @param Product $product
+     * @return array Array of conflicting Url Keys.
+     */
+    public function findUrlKeyConflicts(Product $product): array
```

Comment:
A `Validator` class with a method `findUrlKeyConflicts` probably doesn't make much sense, I would recommend actually changing it to `validateUrlKeyConflicts` and throw the exception here. Make this method a void instead, there is no point to return the URL keys if they won't be used.

Reply:
Also, agree with Myroslav, the generation should happen on controller and be a parameter on this function. If the information from product that we need is just ID, probably just use it as parameter instead of whole product object.

Reply:
I am not sure that rework this to passing UrlKey parameter is a good idea. 

In this case the most of logic will stay in the controller. Because there is a function that is using for UrlPath generation
`$this->productUrlPathGenerator->getUrlPathWithSuffix($product, $store->getId());`,
that uses a product model. The only what will left in the validator is "findAllByData" and the pieace of program that checks found rewrites are not belong to a product.

The other pieace of program generating urlKey, generating urlPathes of stores, searching storeIds of websites (as workaround for performance) will be in the controller.

Please check the last changes and then confirm or deny that it should be reworked like you said before.

Reply:
Thank you, fixed. Validator now throws exception. And method renamed.